### DISCUSSION TITLE
Allow for publish-time overrides to package.json

### DISCUSSION THREAD

**Discussion by drbenschmidt** [npm/feedback#511] [CLI] [answered]
**Problem Statement**

I have a repository setup using workspaces and it uses typescript. In order for testing and development scripts to run without needing to build, I have the `main` `package.json` property pointing to each package's `/src/index.ts` file. Everything from inside the repository is happy as the various tools use some form of typescript plugins to consume those files. However, at publish time, I'd like to ship a rollup artifact rather than referencing the initial typescript file (ex: `"main": "/dist/index.js"` instead).

**Potential Solution**

I erroneously thought that `publishConfig` would allow `package.json` overrides rather than just `config` options specific to publishing - but having something like that for the publish time swapping of `package.json` properties would be something I can see being useful in this scenario.

**Current Workaround**

We can use one of the lifecycle events to have a script modify the `package.json` before packing and publishing.

**Comment by ljharb** [comment]
Usually the way a TS project avoids needing to build for dev is to use `ts-node` - but for node_modules, which should all be transpiled prior to being published, you'd omit the extension in the "main" (omitting the extension is always a best practice anyways), and then ts-node will pick the `.ts` if present, and node will pick the `.js`.

**Reply by drbenschmidt** [reply]
I wasn't aware we could omit the extension, that's good to know! However, I'm not sure this would solve the problem. Internally the tooling all supports transpiling the local typescript, so running tests or linting will work without the need to build the project. That's a behavior I want to retain, but this presents a problem when publishing because it's referencing the typescript files rather than the transpiled ones. Even if I were to omit the extension, the transpiled files go to another directory.

**Reply by ljharb** [reply]
Right - the most common, and safest, scenario here is to sacrifice the local dev experience for the far far more important scenario, published consumption.

**Reply by drbenschmidt** [reply]
> sacrifice the local dev experience

That sounds like we can/should come up with a solution for this to me 😄  The end published package is the highest priority, but I'm not sure we need to neglect the developer experience to get there.

**Reply by ljharb** [reply]
Totally! i think an [RFC](https://github.com/npm/rfcs) for something to solve this would be great!

However, I think it should be on *top of* the status quo here, which is that "main" is always "what gets published" - and provide an opt-in layer for local dev. I'd love to read what you come up with!

**Comment by MylesBorins** [comment, marked as answer]
As mentioned by @ljharb the best next step would be to open an Issue over at the [RFC Repo](https://github.com/npm/rfcs) to kick off a discussion about what this feature could look like with the CLI team
